File: .public-export/torve-public-candidate/androidApp/src/main/java/is/xyz/mpv/MPVLib.java
package is.xyz.mpv;

import android.content.Context;
import android.graphics.Bitmap;
import android.view.Surface;
import androidx.annotation.NonNull;
import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("unused")
public final class MPVLib {

  private static final List<EventObserver> observers = new ArrayList<>();
  private static final List<LogObserver> logObservers = new ArrayList<>();

  private MPVLib() {}

  public static native void create(Context context);
  public static native void init();
  public static native void destroy();

  public static native void attachSurface(Surface surface);
  public static native void detachSurface();

  public static native void command(@NonNull String[] cmd);
  public static native int setOptionString(String name, String value);
  public static native Bitmap grabThumbnail(int dimension);

  public static native void setPropertyBoolean(@NonNull String name, @NonNull Boolean value);
  public static native void setPropertyInt(@NonNull String name, @NonNull Integer value);
  public static native void setPropertyDouble(@NonNull String name, @NonNull Double value);
  public static native void setPropertyString(@NonNull String name, @NonNull String value);

  public static native Boolean getPropertyBoolean(@NonNull String name);
  public static native Integer getPropertyInt(@NonNull String name);
  public static native Double getPropertyDouble(@NonNull String name);
  public static native String getPropertyString(@NonNull String name);

  public static native void observeProperty(@NonNull String name, int format);

  public static void addObserver(EventObserver observer) {
    synchronized (observers) {
      observers.add(observer);
    }
  }

  public static void removeObserver(EventObserver observer) {
    synchronized (observers) {
      observers.remove(observer);
    }
  }

  public static void addLogObserver(LogObserver observer) {
    synchronized (logObservers) {
      logObservers.add(observer);
    }
  }

  public static void removeLogObserver(LogObserver observer) {
    synchronized (logObservers) {
      logObservers.remove(observer);
    }
  }

  public static void eventProperty(String property, long value) {
    synchronized (observers) {
      for (EventObserver observer : observers) {
        observer.eventProperty(property, value);
      }
    }
    com.torve.android.player.MPVLib.dispatchEventProperty(property, value);
  }

  public static void eventProperty(String property, boolean value) {
    synchronized (observers) {
      for (EventObserver observer : observers) {
        observer.eventProperty(property, value);
      }
    }
    com.torve.android.player.MPVLib.dispatchEventProperty(property, value);
  }

  public static void eventProperty(String property, double value) {
    synchronized (observers) {
      for (EventObserver observer : observers) {
        observer.eventProperty(property, value);
      }
    }
    com.torve.android.player.MPVLib.dispatchEventProperty(property, value);
  }

  public static void eventProperty(String property, String value) {
    synchronized (observers) {
      for (EventObserver observer : observers) {
        observer.eventProperty(property, value);
      }
    }
    com.torve.android.player.MPVLib.dispatchEventProperty(property, value);
  }

  public static void eventProperty(String property) {
    synchronized (observers) {
      for (EventObserver observer : observers) {
        observer.eventProperty(property);
      }
    }
    com.torve.android.player.MPVLib.dispatchEventProperty(property, null);
  }

  public static void event(int eventId) {
    synchronized (observers) {
      for (EventObserver observer : observers) {
        observer.event(eventId);
      }
    }
    com.torve.android.player.MPVLib.dispatchEvent(eventId);
  }

  public static void logMessage(String prefix, int level, String text) {
    synchronized (logObservers) {
      for (LogObserver observer : logObservers) {
        observer.logMessage(prefix, level, text);
      }
    }
    if (com.torve.android.BuildConfig.DEBUG) {
      android.util.Log.d(
          "UpstreamMPVLib",
          prefix + "[" + level + "]: " +
              com.torve.domain.diagnostics.DiagnosticsRedactor.INSTANCE.redact(text));
    }
  }

  public interface EventObserver {
    void eventProperty(@NonNull String property);
    void eventProperty(@NonNull String property, long value);
    void eventProperty(@NonNull String property, boolean value);
    void eventProperty(@NonNull String property, @NonNull String value);
    void eventProperty(@NonNull String property, double value);
    void event(int eventId);
  }

  public interface LogObserver {
    void logMessage(@NonNull String prefix, int level, @NonNull String text);
  }
}

File: androidApp/src/main/kotlin/com/torve/android/player/MPVLib.kt
package com.torve.android.player

import android.content.Context
import android.view.Surface
import `is`.xyz.mpv.MPVLib as UpstreamMPVLib

/**
 * JNI bridge to libmpv. Native methods map to mpv's client API.
 * Requires libmpv .so files in jniLibs/ (arm64-v8a, armeabi-v7a, x86_64).
 */
object MPVLib {

    private var isLoaded = false
    private var isInitialized = false

    /**
     * Try to load native libraries. Returns true if successful.
     */
    fun tryLoad(): Boolean {
        if (isLoaded) return true
        return try {
            System.loadLibrary("c++_shared")
            System.loadLibrary("avutil")
            System.loadLibrary("swresample")
            System.loadLibrary("swscale")
            System.loadLibrary("avcodec")
            System.loadLibrary("avformat")
            System.loadLibrary("avfilter")
            System.loadLibrary("avdevice")
            System.loadLibrary("mpv")
            System.loadLibrary("player")
            isLoaded = true
            true
        } catch (_: UnsatisfiedLinkError) {
            false
        }
    }

    fun isAvailable(): Boolean = isLoaded
    fun isInitialized(): Boolean = isLoaded && isInitialized

    // --- Initialization ---

    fun create(context: Context) = UpstreamMPVLib.create(context)
    fun init() {
        UpstreamMPVLib.init()
        isInitialized = true
    }

    fun destroy() {
        if (!isInitialized()) return
        UpstreamMPVLib.destroy()
        isInitialized = false
    }

    fun setOptionString(name: String, value: String): Int = UpstreamMPVLib.setOptionString(name, value)

    // --- Surface ---

    fun attachSurface(surface: Surface) {
        if (!isInitialized()) return
        UpstreamMPVLib.attachSurface(surface)
    }

    fun detachSurface() {
        if (!isInitialized()) return
        UpstreamMPVLib.detachSurface()
    }

    // --- Playback control ---

    fun command(cmd: Array<String>) = UpstreamMPVLib.command(cmd)

    fun loadFile(url: String) {
        command(arrayOf("loadfile", url))
    }

    fun play() {
        setPropertyBoolean("pause", false)
    }

    fun pause() {
        setPropertyBoolean("pause", true)
    }

    fun stop() {
        command(arrayOf("stop"))
    }

    fun seek(positionSec: Double) {
        command(arrayOf("seek", positionSec.toString(), "absolute"))
    }

    fun seekRelative(deltaSec: Double) {
        command(arrayOf("seek", deltaSec.toString(), "relative"))
    }

    // --- Properties ---

    fun setPropertyBoolean(name: String, value: Boolean) = UpstreamMPVLib.setPropertyBoolean(name, value)
    fun setPropertyInt(name: String, value: Int) = UpstreamMPVLib.setPropertyInt(name, value)
    fun setPropertyDouble(name: String, value: Double) = UpstreamMPVLib.setPropertyDouble(name, value)
    fun setPropertyString(name: String, value: String) = UpstreamMPVLib.setPropertyString(name, value)

    fun getPropertyBoolean(name: String): Boolean = UpstreamMPVLib.getPropertyBoolean(name)
    fun getPropertyInt(name: String): Int = UpstreamMPVLib.getPropertyInt(name)
    fun getPropertyDouble(name: String): Double = UpstreamMPVLib.getPropertyDouble(name)
    fun getPropertyString(name: String): String? = UpstreamMPVLib.getPropertyString(name)

    // --- Observe properties (callbacks dispatched to EventThread) ---

    fun observeProperty(name: String, format: Int) = UpstreamMPVLib.observeProperty(name, format)

    // MPV property format constants
    const val MPV_FORMAT_NONE = 0
    const val MPV_FORMAT_STRING = 1
    const val MPV_FORMAT_FLAG = 3
    const val MPV_FORMAT_INT64 = 4
    const val MPV_FORMAT_DOUBLE = 5

    // --- Track info ---

    data class Track(
        val id: Int,
        val type: String, // "audio", "sub", "video"
        val title: String?,
        val language: String?,
        val codec: String?,
        val isDefault: Boolean,
        val isSelected: Boolean,
    )

    fun getTracks(): List<Track> {
        if (!isLoaded) return emptyList()
        val count = try { getPropertyInt("track-list/count") } catch (_: Exception) { 0 }
        val selectedAudioTrackId = try {
            getPropertyInt("aid")
        } catch (_: Exception) {
            getPropertyString("aid")?.toIntOrNull() ?: -1
        }
        val tracks = mutableListOf<Track>()
        for (i in 0 until count) {
            try {
                val type = getPropertyString("track-list/$i/type") ?: continue
                val trackId = getPropertyInt("track-list/$i/id")
                val selectedByTrackList = try { getPropertyBoolean("track-list/$i/selected") } catch (_: Exception) { false }
                tracks.add(
                    Track(
                        id = trackId,
                        type = type,
                        title = getPropertyString("track-list/$i/title"),
                        language = getPropertyString("track-list/$i/lang"),
                        codec = getPropertyString("track-list/$i/codec"),
                        isDefault = try { getPropertyBoolean("track-list/$i/default") } catch (_: Exception) { false },
                        isSelected = selectedByTrackList || (type == "audio" && trackId == selectedAudioTrackId),
                    ),
                )
            } catch (_: Exception) {
                // Skip malformed track entries
            }
        }
        return tracks
    }

    fun selectAudioTrack(trackId: Int) {
        setPropertyString("aid", trackId.toString())
    }

    fun selectSubtitleTrack(trackId: Int) {
        setPropertyString("sid", trackId.toString())
    }

    fun disableSubtitles() {
        setPropertyString("sid", "no")
    }

    // --- Event listener ---

    interface EventObserver {
        fun onPropertyChange(property: String, value: Any?)
        fun onEvent(eventId: Int)
    }

    private val observers = mutableListOf<EventObserver>()

    fun addObserver(observer: EventObserver) {
        observers.add(observer)
    }

    fun removeObserver(observer: EventObserver) {
        observers.remove(observer)
    }

    // Called from JNI
    @JvmStatic
    fun dispatchEventProperty(property: String, value: Any?) {
        observers.forEach { it.onPropertyChange(property, value) }
    }

    @JvmStatic
    fun dispatchEvent(eventId: Int) {
        observers.forEach { it.onEvent(eventId) }
    }
}
